exampleCourse/questions/workshop/Lesson4_example2/server.py

```diff
@@ -25,3 +25,15 @@ def grade(data):
         data["score"] = 1
     else:
         data["score"] = 0
+
+
+def test(data):
+    if data["test_type"] != "correct":
+        raise NotImplementedError
```

Comment:
Is this function only called if the test type is correct? Does this not raise an error?

Reply:
This does raise an error if triggered via the UI -- the "incorrect" and "invalid" tests fail. This is an incremental improvement of the existing behavior -- which is to simply crash. I am not a fan of requiring instructors to implement all three -- and this work is to support #11837

Reply:
Re: #11837, historically we've been deliberate about which questions we include in our automated tests. We should consider the same here, if only in the form of an exclusion list.

The example course serves an interesting role: it contains template questions (high-quality, gold-standard), examples of elements, questions from workshops, and more. To the extent that people are looking to any of these questions for best practices, I think we should either a) implement real tests, or b) not implement tests at all and exclude these from the automated tests you're working on.

Reply:
In [ded45dd](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/ded45dd647d18e45cfc9514fd301303c65e58f87) I updated the elements so they still can generate invalid inputs for some scenarios. Then, I'll update these to support generating both "correct" and "incorrect" inputs.

Reply:
Updated more in [08ec836](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/08ec83674631cfbf692eb050cfe13573f9c96c46).